### PR TITLE
Distance calculation functions for doubles - L2 SSE

### PR DESCRIPTION
**Pull request by meiravgri**
In this PR we added fp64 support to **euclidean distance** calculation functions using **SSE optimization**

### CODE REVIEW

File: tests/unit/test_spaces.cpp [hunk 1]
```diff
@@ -118,6 +118,27 @@ TEST_F(SpacesTest, l2_17) {
...
+    for (size_t i = 0; i < dim; i++) {
+        v[i] = (double)i;
+    }
+
+    double baseline = FP64_L2Sqr(v, v, dim);
+    switch (optimization) {
+    case ARCH_OPT_AVX512: // TODO: add comparison when AVX and AVX512 is implemented
+    case ARCH_OPT_AVX:
+    case ARCH_OPT_SSE:
+    default:
```
Comment (alonre24) [inline tests/unit/test_spaces.cpp:135]: What about the case of no optimization?

File: tests/unit/test_spaces.cpp [hunk 2]
```diff
@@ -118,6 +118,27 @@ TEST_F(SpacesTest, l2_17) {
...
+        ASSERT_EQ(baseline, FP64_L2SqrSIMD8ExtResiduals_SSE(v, v, dim));
+        break;
+        ;
```
Comment (alonre24) [inline tests/unit/test_spaces.cpp:138]: redundant semi colon ;)

File: src/VecSim/spaces/spaces.cpp [hunk 3]
```diff
@@ -4,26 +4,43 @@
...
+    }
+    return ret_score;
+}
+
+CalculationGuideline FP64_GetCalculationGuideline(size_t dim) {
+
+    CalculationGuideline ret_score = NO_OPTIMIZATION;
+
+    if (dim % 8 == 0) {
+        ret_score = ITER_512_BITS;
```
Comment (alonre24) [inline src/VecSim/spaces/spaces.cpp:37]: Why there is no coverage for this?

**Review by alonre24 — COMMENTED** [review]
Looks good!! Only had small questions regarding tests and coverage - (applies to all tests in test_spaces file)

**Review by alonre24 — APPROVED** [review]

File: src/VecSim/spaces/spaces.cpp [hunk 4]
```diff
@@ -4,26 +4,43 @@
...
+/*** Defined in spaces.h
+ //optimization are defined according to
+ // num_bits_per_iteration / bits_count_per_element
     NO_OPTIMIZATION = 0,
-    Ext16 = 1,          // dim % 16 == 0
-    Ext4 = 2,           // dim % 4 == 0
-    ExtResiduals16 = 3, // dim > 16 && dim % 16 < 4
-    ExtResiduals4 = 4,  // dim > 4
-};
+    ITER_512_BITS = 1,          // FP32 -> dim % 16 == 0, FP64 -> dim % 8 == 0
```
Comment (DvirDukhan) [inline src/VecSim/spaces/spaces.cpp:11]:
```suggestion
   512_BITS = 1,          // FP32 -> dim % 16 == 0, FP64 -> dim % 8 == 0
```

File: src/VecSim/spaces/spaces.h [hunk 5]
```diff
@@ -18,15 +18,23 @@ using dist_func_t = RET_TYPE (*)(const void *, const void *, size_t);
 
 enum CalculationGuideline {
     NO_OPTIMIZATION = 0,
-    Ext16 = 1,          // dim % 16 == 0
-    Ext4 = 2,           // dim % 4 == 0
-    ExtResiduals16 = 3, // dim > 16 && dim % 16 < 4
-    ExtResiduals4 = 4,  // dim > 4
+    ITER_512_BITS = 1, // FP32 -> dim % 16 == 0, FP64 -> dim % 8 == 0
```
Comment (DvirDukhan) [inline src/VecSim/spaces/spaces.h:21]:
```suggestion
    512_BITS = 1, // FP32 -> dim % 16 == 0, FP64 -> dim % 8 == 0
```

File: src/VecSim/spaces/spaces.h [hunk 6]
```diff
@@ -18,15 +18,23 @@ using dist_func_t = RET_TYPE (*)(const void *, const void *, size_t);
...
+    ITER_128_BITS = 2, // FP32 -> dim % 4 == 0, FP64 -> dim % 2 == 0
```
Comment (DvirDukhan) [inline src/VecSim/spaces/spaces.h:22]:
```suggestion
    128_BITS = 2, // FP32 -> dim % 4 == 0, FP64 -> dim % 2 == 0
```

File: src/VecSim/spaces/spaces.h [hunk 7]
```diff
@@ -18,15 +18,23 @@ using dist_func_t = RET_TYPE (*)(const void *, const void *, size_t);
...
+    ITER_512_BITS_RESIDUALS =
```
Comment (DvirDukhan) [inline src/VecSim/spaces/spaces.h:23]:
```suggestion
    512_BITS_RESIDUALS =
```

File: src/VecSim/spaces/spaces.h [hunk 8]
```diff
@@ -18,15 +18,23 @@ using dist_func_t = RET_TYPE (*)(const void *, const void *, size_t);
...
 
 void SetDistFunc(VecSimMetric metric, size_t dim, dist_func_t<float> *index_dist_func);
 void SetDistFunc(VecSimMetric metric, size_t dim, dist_func_t<double> *index_dist_func);
 
+// Useful numbers
+const unsigned int LOG_BASE2_16 = 4;
+const unsigned int LOG_BASE2_8 = 3;
+const unsigned int LOG_BASE2_4 = 2;
+const unsigned int LOG_BASE2_2 = 1;
+
```
Comment (DvirDukhan) [inline src/VecSim/spaces/spaces.h:39]: not in use

File: src/VecSim/spaces/spaces.h [hunk 9]
```diff
@@ -18,15 +18,23 @@ using dist_func_t = RET_TYPE (*)(const void *, const void *, size_t);
...
     NO_OPTIMIZATION = 0,
-    Ext16 = 1,          // dim % 16 == 0
-    Ext4 = 2,           // dim % 4 == 0
-    ExtResiduals16 = 3, // dim > 16 && dim % 16 < 4
-    ExtResiduals4 = 4,  // dim > 4
+    ITER_512_BITS = 1, // FP32 -> dim % 16 == 0, FP64 -> dim % 8 == 0
+    ITER_128_BITS = 2, // FP32 -> dim % 4 == 0, FP64 -> dim % 2 == 0
+    ITER_512_BITS_RESIDUALS =
+        3, // FP32 ->  dim > 16 && dim % 16 < 4, FP64 -> dim > 8 && dim % 8 < 2,
+    ITER_128_BITS_RESIDUALS = 4, // FP32 ->dim > 4, FP64 -> dim > 2
```
Comment (DvirDukhan) [inline src/VecSim/spaces/spaces.h:25]:
```suggestion
    128_BITS_RESIDUALS = 4, // FP32 ->dim > 4, FP64 -> dim > 2
```

File: src/VecSim/spaces/L2/L2_SSE_FP64.cpp [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    // TmpRes must be 16 bytes aligned
+    double PORTABLE_ALIGN16 TmpRes[2];
+    _mm_store_pd(TmpRes, sum);
+    return TmpRes[0] + TmpRes[1];
+}
+
+double FP64_L2SqrSIMD8ExtResiduals_SSE(const void *pVect1v, const void *pVect2v, size_t qty) {
+    // Calc how many doubles we can calc using 512 bits iterations.
```
Comment (DvirDukhan) [inline src/VecSim/spaces/L2/L2_SSE_FP64.cpp:52]:
```suggestion
    // Calculate how many doubles we can calculate using 512 bits iterations.
```

File: src/VecSim/spaces/L2/L2_SSE_FP64.cpp [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    size_t qty8 = qty >> 3 << 3;
+    double res = FP64_L2SqrSIMD8Ext_SSE(pVect1v, pVect2v, qty8);
+    double *pVect1 = (double *)pVect1v + qty8;
+    double *pVect2 = (double *)pVect2v + qty8;
+
+    // Calc the rest using a brute force function
```
Comment (DvirDukhan) [inline src/VecSim/spaces/L2/L2_SSE_FP64.cpp:58]:
```suggestion
    // Calculate the rest using the basic function
```

File: src/VecSim/spaces/L2/L2_SSE_FP64.cpp [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    double res_tail = FP64_L2Sqr(pVect1, pVect2, qty_left);
+
+    return (res + res_tail);
+}
+
+double FP64_L2SqrSIMD2Ext_SSE(const void *pVect1v, const void *pVect2v, size_t qty) {
+    double *pVect1 = (double *)pVect1v;
+    double *pVect2 = (double *)pVect2v;
+
+    // Calc how many doubles we can calc using 512 bits iterations.
```
Comment (DvirDukhan) [inline src/VecSim/spaces/L2/L2_SSE_FP64.cpp:69]:
```suggestion
    // Calculate how many doubles we can calculate using 512 bits iterations.
```

**Review by DvirDukhan — CHANGES_REQUESTED** [review]

File: src/VecSim/spaces/spaces.h [hunk 13]
```diff
@@ -18,23 +18,16 @@ using dist_func_t = RET_TYPE (*)(const void *, const void *, size_t);
...
     NO_OPTIMIZATION = 0,
-    ITER_512_BITS = 1, // FP32 -> dim % 16 == 0, FP64 -> dim % 8 == 0
-    ITER_128_BITS = 2, // FP32 -> dim % 4 == 0, FP64 -> dim % 2 == 0
-    ITER_512_BITS_RESIDUALS =
+    SPLIT_TO_512_BITS = 1,     // FP32 -> dim % 16 == 0, FP64 -> dim % 8 == 0
+    SPLIT_TO_512_128_BITS = 2, // FP32 -> dim % 4 == 0, FP64 -> dim % 2 == 0
+    SPLIT_TO_512_BITS_WITH_RESIDUALS =
         3, // FP32 ->  dim > 16 && dim % 16 < 4, FP64 -> dim > 8 && dim % 8 < 2,
-    ITER_128_BITS_RESIDUALS = 4, // FP32 ->dim > 4, FP64 -> dim > 2
+    SPLIT_TO_128_BITS_WITH_RESIDUALS = 4, // FP32 ->dim > 4, FP64 -> dim > 2
```
Comment (DvirDukhan) [inline src/VecSim/spaces/spaces.h:25]:
```suggestion
    SPLIT_TO_512_128_BITS_WITH_RESIDUALS = 4, // FP32 ->dim > 4, FP64 -> dim > 2
```

**Pull request merged**